examples/mulmat-tune/README.md

```diff
@@ -0,0 +1,574 @@
+token size >= 32. `ggml_compute_forward_mul_mat_use_blas` checks:
+- data of `src0` and `src1` are continuous
+- M >=32 &&  >= 32 && K >= 32.
+It's obvious that this simple rule is not accurate at all. There are several TODOs
+that mentioned this problem.
+
+I have been focused on the `threading` problem(s) since Apr this year. I dropped
+two simple pull requests because they are not solutions but noises. In the second
+pull request, @gerganov hinted me the `1-thread blas` problem, so I followed this
+direction since then. My benchmarks on 7B/13B and all Qx_x shows that: for all
+N/K (e.g, 4096x4096), Accelerate always run faster than pure GPU at `M = 32`,
+the difference may vary up to `30%` as of 1 thread. I will show you image for this.
+
+When I observed that the de-quantization takes about half of total time, I think
+it's a good chance to start from here. At first I implemented the new threading
+framework that supports `wait/notify`, subtle and fragile to dead lock. I'm happy
+that it works. I had tried to bench online by comparing CPU/GPU time, finally I
+replaced that with offline bench. To explicitly control details (how to parallel,
+when to wait, how to select the best executing plan), I had to define task config,
+task profiles. Finally after over a month busy coding I got the demo solution.
+
+New CUDA and GPU codes are keep changing in recent days, new rules for when and
+how to run `mul_mat` were introduced. I often had trouble in merging/rebasing
+with master, let alone understand the details. For example:
+
+- To test ClBlast on my device, I managed to solve the OOM problem with smaller
+  matrices in multi-passes. Finally unable to catchup with recent change.
+- I periodically refactor codes to keep up with `enum ggml_backend`. Also, I just
+  noticed [multi GPU support in #1670](https://github.com/ggerganov/llama.cpp/pull/1607), looks like this PR will change
+  `enum ggml_backend` again. Although I have'nt read all codes in this PR and
+  must not able to understand all even if I did, but I like one of the changes
+  that moving `enum ggml_task_type` and `struct ggml_task_params`.
+
+In the following sections, all bench data will will be mentioned are generated
+on MacBook Pro 2018 with 32 GB 2400 MHz DDR4 memory, 2.6 GHz 6-Core Intel Core
+i7-8850H @2.60GHz, Intel UHD Graphics 630 1536 MB.
+
+When you read the following sections, please note that current incoming codes are
+based on this commit and fragile to forced pushing:
+
+```
+commit 3b126f654fceb4f5f195a1c2e825bb18e101188c
+Author: Johannes Gäßler <johannesg@5d6.de>
+Date:   Sun May 28 21:01:02 2023 +0200
+```
+
+## Solution and Result
+
+In current `master` branch, the `mul_mat` codes run in several implicit profiles.
+
+- pure cpu: INIT: very fast, COMPUTE: the computation time is proportional to M.
+- CUDA/CL: COMPUTE: de-quantization and mul_mat in GPU.
+- Accelerate/OpenBLAS/BLIS: COMPUTE: de-quantization in CPU, mul_mat in GPU.
```

Comment:
Not sure about accelerate, but OpenBLAS and BLIS are pure CPU libraries. Their mul_mat happens on CPU.